Redesign guest pages for better accessibility

CarlSchwan opened the pull request:
Screenshots:
![login](https://user-images.githubusercontent.com/23653902/179971351-038bcd36-be3e-493d-8c3e-2004fa052f5d.png)

![image](https://user-images.githubusercontent.com/23653902/179678157-53bb70a4-1d94-4e12-8719-a0baf86a51a5.png)

![image](https://user-images.githubusercontent.com/23653902/179681268-160a54ea-8bf9-4e5d-aa26-eb593ee50ae2.png)

![image](https://user-images.githubusercontent.com/23653902/179681763-f1e413f0-2c83-4ffe-9993-6a5a5fda7290.png)

![image](https://user-images.githubusercontent.com/23653902/179681443-17433a6f-3b15-4252-a56d-d446c9068b79.png)

![image](https://user-images.githubusercontent.com/23653902/179680065-30baee88-1b18-4b0e-816a-ab567e55355c.png)

![install1](https://user-images.githubusercontent.com/23653902/179971474-2cdf1a81-680f-469b-9655-a90bcc432825.png)

![install1b](https://user-images.githubusercontent.com/23653902/179971447-aa662d43-90ed-48d3-9857-e39539e8c4f2.png)

![install2](https://user-images.githubusercontent.com/23653902/179971571-de270773-103c-48ad-a33c-ce41602fdeae.png)

![install3](https://user-images.githubusercontent.com/23653902/179971602-857d97a4-30ba-4c55-ae3a-98d23ca15136.png)

![image](https://user-images.githubusercontent.com/23653902/179979809-81ec8edb-5f0d-428b-abb9-c75944fadd0e.png)

TODO:

- [x] Adapt installer to use the same style
- [x] Adapt updater to use the same style
- [x] Adapt 2fa pages to use the same style
- [x] Find other page using this style and adapt them

CarlSchwan committed:
Redesign guest pages for better accessibility

- Use white box and put content on it
- Improve focus indicator

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

jancborchardt commented:
Looks great! :) 2 things:
- For the first screenshot, "Access forbidden" should be marked up as a heading, right?
- These are all wide boxes, do you have a screenshot of the login view (with the username & password boxes) as well?

CarlSchwan commented:
> For the first screenshot, "Access forbidden" should be marked up as a heading, right?

Right this was a preexisting issue, fixed now ;)

> These are all wide boxes, do you have a screenshot of the login view (with the username & password boxes) as well?

I updated all the box to not be wide anymore by default unless specified otherwise like on the exception page. Screenshots of the new states are updated

jancborchardt reviewed (changes requested):
Very nice, looks so good now! :) Only 2 things:
- [x] There is focus feedback on the username/password input field, but they don’t have a default border. Either using the component, or `border: 2px solid var(--color-border)` would look nice.
- [x] From the screenshots, all the scenarios except the internal server error look better with the narrow style – so I’d say let’s change the "Connect to your account" and "Account access" to narrow style. :)

nimishavijay reviewed (approved):
All the screenshots look great! :) 
Not sure if this is relevant to this PR but we should change everything that was full opacity black `#000000` to `--color-main-text`, it's especially visible in the screenshot with the app icons.

nickvergessen commented on core/src/components/setup/RecommendedApps.vue:
```diff
@@ -20,16 +20,16 @@
 		<h2>{{ t('core', 'Recommended apps') }}</h2>
 		<p v-if="loadingApps" class="loading text-center">
-			{{ t('core', 'Loading apps …') }}
+			{{ t('core', 'Loading apps…') }}
```
Since it's not shortcutting the word "apps" the space is required

nickvergessen commented on core/src/components/setup/RecommendedApps.vue:
```diff
@@ -20,16 +20,16 @@
 		</p>
 		<p v-else-if="installingApps" class="text-center">
-			{{ t('core', 'Installing apps …') }}
+			{{ t('core', 'Installing apps…') }}
```
Since it's not shortcutting the word "apps" the space is required

CarlSchwan replied:
reverted, but this feels weird. Is this convention documented somewhere?

nickvergessen replied:
@rakekniven @jancborchardt anyone has a link?

nickvergessen commented on core/src/components/setup/RecommendedApps.vue:
```diff
@@ -53,12 +53,21 @@
-			<a :href="defaultPageUrl">{{ t('core', 'Cancel') }}</a>
-		</p>
+		<div class="dialog-row">
+			<Button type="tertiary-no-background"
```
```suggestion
			<Button type="tertiary"
```

No background should only be used inside of inputs

nickvergessen replied:
This is still not resolved?

nickvergessen commented on core/src/components/setup/RecommendedApps.vue:
```diff
@@ -215,7 +228,9 @@ p {
 		height: 50px;
 		width: 50px;
-		filter: invert(1);
+		@media (prefers-color-scheme: dark) {
+			filter: invert(1);
+		}
```
```suggestion
		filter: var(--background-invert-if-dark);
```

nickvergessen replied:
```suggestion
		filter: var(--background-invert-if-dark);
```

To respect the user preference in case they overwrite the browser prefers setting?
Or does it not work on the install/update page?

nickvergessen commented on core/src/views/Login.vue:
```diff
@@ -20,87 +20,106 @@
+			<transition name="fade" mode="out-in">
+				<div class="warning">
+					{{ t('core', 'Login form is disabled.') }}<br>
+					<small>{{ t('core', 'Please contact your administrator.') }}
+					</small>
```
Either linebreak both or none 😿

nickvergessen commented on core/templates/installation.php:
```diff
@@ -161,7 +163,7 @@
 	<div class="icon-loading-dark float-spinner">&nbsp;</div>
 
-	<div class="buttons"><input type="submit" class="primary" value="<?php p($l->t('Install')); ?>" data-finishing="<?php p($l->t('Installing …')); ?>"></div>
+	<div class="buttons"><input type="submit" class="primary" value="<?php p($l->t('Install')); ?>" data-finishing="<?php p($l->t('Installing…')); ?>"></div>
```
Since it's not shortcutting the word "installing" the space is required

nickvergessen commented on core/templates/layout.guest.php:
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html class="ng-csp" data-placeholder-focus="false" lang="<?php p($_['language']); ?>" data-locale="<?php p($_['locale']); ?>" >
+<html class="ng-csp nc-<?php p($_['major_version']); ?>" data-placeholder-focus="false" lang="<?php p($_['language']); ?>" data-locale="<?php p($_['locale']); ?>" >
```
I don't think we should start this.

nickvergessen commented on core/templates/layout.public.php:
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html class="ng-csp" data-placeholder-focus="false" lang="<?php p($_['language']); ?>" data-locale="<?php p($_['locale']); ?>" >
+<html class="ng-csp nc-<?php p($_['major_version']); ?>" data-placeholder-focus="false" lang="<?php p($_['language']); ?>" data-locale="<?php p($_['locale']); ?>" >
```
Same here. This is just going to bring us to a hell....

nickvergessen commented on core/templates/loginflowv2/authpicker.php:
```diff
@@ -36,17 +36,17 @@
 
 	<p id="redirect-link">
-		<a href="<?php p($urlGenerator->linkToRouteAbsolute('core.ClientFlowLoginV2.grantPage', ['stateToken' => $_['stateToken'], 'user' => $_['user']])) ?>">
+		<a href="<?php p($urlGenerator->linkToRouteAbsolute('core.ClientFlowLoginV2.grantPage', ['stateToken' => $_['stateToken'], 'user' => $_['user']])) ?>"  tabindex="-1">
```
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

> A negative value is useful when you have off-screen content that appears on a specific event. The user won't be able to focus any element with a negative tabindex using the keyboard, but a script can do so by calling the focus() [method](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus).

doesn't sound like the correct usecase?

CarlSchwan replied:
I replaced the `<a>` with a `<form>` and this should do the trick without having two element tab focusable but doing the same

nickvergessen commented on lib/private/TemplateLayout.php:
```diff
@@ -194,6 +194,7 @@ public function __construct($renderAs, $appId = '') {
 		$lang = str_replace('_', '-', $lang);
 		$this->assign('language', $lang);
 		$this->assign('locale', $locale);
+		$this->assign('major_version', \OCP\Util::getVersion()[0]);
```
😿

nickvergessen commented on core/css/guest.css:
```diff
@@ -947,3 +861,79 @@ label.infield,
+	height: 44px;
+}
+
+.mt-3, .my-3 {
+	margin-top: 1rem;
+}
+
+.mb-3, .my-3 {
+	margin-top: 1rem;
+}
```
those names -.-

Pytal commented on core/css/guest.css:
```diff
@@ -947,3 +861,79 @@ label.infield,
+	display: inline-block;
+}
+
+.text-left {
+	text-align: left;
+}
```
Are we introducing in-house Tailwind CSS?

nickvergessen commented on core/src/components/login/LoginButton.vue:
```diff
@@ -59,36 +64,13 @@ export default {
-		applied to the button instead. */
-		pointer-events: none;
+<style lang="scss">
+	.button-inverted {
+		&.button-vue--icon-and-text {
+			padding: 0 4px 0 16px;
+		}
+		.button-vue__wrapper {
+			flex-direction: row-reverse;
+		}
 	}
```
🙈  I was about to find where this breaks, but you do it intentionally.
I think it would be time to follow the common pattern and have a useful icon (not `->`) or no icon.

Also the icon is now on all buttons (social login, registration, ...) which looks awkward

nickvergessen commented on core/src/views/Login.vue:
```diff
@@ -20,87 +20,107 @@
+		<div v-if="!hideLoginForm || directLogin">
+			<transition name="fade" mode="out-in">
+				<div v-if="!passwordlessLogin && !resetPassword && resetPasswordTarget === ''"
+					key="login">
```
invalid attribute?!

nickvergessen commented on core/src/views/Login.vue:
```diff
@@ -20,87 +20,107 @@
+					key="reset"
+					class="login-additional">
+					<div class="lost-password-container">
+						<ResetPassword v-if="resetPassword"
+							:username.sync="user"
+							:reset-password-link="resetPasswordLink"
+							@abort="resetPassword = false" />
```
Text is not readable in dark mode:
![Bildschirmfoto vom 2022-07-22 13-10-27](https://user-images.githubusercontent.com/213943/180427356-774fd254-9bd4-4907-9c48-05b9598f3796.png)

nickvergessen commented on core/src/views/Login.vue:
```diff
@@ -20,87 +20,107 @@
+				role="link"
+				:href="alternativeLogin.href">
+				{{ alternativeLogin.name }}
+				<template #icon>
+					<ArrowRight />
+				</template>
```
```suggestion
```

No duplicate icons please

nickvergessen commented on core/src/views/Login.vue:
```diff
@@ -20,87 +20,107 @@
+					<a v-if="canResetPassword && resetPasswordLink !== ''"
+						id="lost-password"
+						:href="resetPasswordLink">
+						{{ t('core', 'Forgot password?') }}
```
TExt is not readable on dark mode

![Bildschirmfoto vom 2022-07-22 13-10-27](https://user-images.githubusercontent.com/213943/180429095-c2c3c781-e572-46cd-bc7a-71661e4f2fbb.png)

CarlSchwan committed:
Fix caldav invitation related public pages

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

AndyScherzinger commented:
@jancborchardt looking at the screenshot
![install3](https://user-images.githubusercontent.com/23653902/179971602-857d97a4-30ba-4c55-ae3a-98d23ca15136.png)

shouldn't "Cancel" rather be "Skip" ?

CarlSchwan commented:
> @jancborchardt looking at the screenshot ![install3](https://user-images.githubusercontent.com/23653902/179971602-857d97a4-30ba-4c55-ae3a-98d23ca15136.png)
> 
> shouldn't "Cancel" rather be "Skip" ?

It should, good catch

AndyScherzinger commented:
Thanks for caring and for taking care @CarlSchwan 💙

CarlSchwan commented:
> Thanks for caring and for taking care @CarlSchwan blue_heart

Fixed :)

I would say this is good to go now. There might be some apps that need to be adapted after this change but I for the moment checked most of them and everything seems to still work fine

nickvergessen commented on core/src/components/setup/RecommendedApps.vue:
```diff
@@ -53,12 +53,21 @@
+				role="link"
+				href="defaultPageUrl"
+				@click="goTo(defaultPageUrl)">
+				{{ t('core', 'Cancel') }}
```
```suggestion
				{{ t('core', 'Skip') }}
```

nickvergessen commented:
> It should, good catch

forgot to push? Or Push failed?

CarlSchwan commented:
> > It should, good catch
> 
> forgot to push? Or Push failed?

forgot to push :(

CarlSchwan committed:
Fix login flow

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

nickvergessen commented on core/templates/loginflow/authpicker.php:
```diff
@@ -36,17 +36,17 @@
-		<a href="<?php p($urlGenerator->linkToRoute('core.ClientFlowLogin.grantPage', ['stateToken' => $_['stateToken'], 'clientIdentifier' => $_['clientIdentifier'], 'oauthState' => $_['oauthState'], 'user' => $_['user'], 'direct' => $_['direct']])) ?>">
+		<form id="login-form" action="<?php p($urlGenerator->linkToRoute('core.ClientFlowLogin.grantPage', ['stateToken' => $_['stateToken'], 'clientIdentifier' => $_['clientIdentifier'], 'oauthState' => $_['oauthState'], 'user' => $_['user'], 'direct' => $_['direct']])) ?>" method="get">
 			<input type="submit" class="login primary icon-confirm-white" value="<?php p($l->t('Log in')) ?>">
 		</a>
```
```suggestion
		</from>
```

jancborchardt commented:
@CarlSchwan this looks so damn cool, and is much more accessible on top of it. Great work! ✨
